capella_ros_tools/exporter.py

```diff
@@ -2,83 +2,604 @@
 import pathlib
 import re
+import typing as t
+from html import parser
 
+import capellambse
+import jinja2
+import pydantic
+import yaml
 from capellambse.metamodel import information
 
-from capella_ros_tools import data_model
-
 from . import logger
 
+PACKAGE_XML = "package.xml"
+C_MAKE_LISTS_TXT = "CMakeLists.txt"
+GENERIC_PKG_NAME = "generic"
+UNKNOWN_TYPE = "unknown"
+DEFAULT_ENUM_TYPE = "int32"
+PACKAGE_PATH = "export_templates"
+ROS_TYPES = [
+    "bool",
+    "byte",
+    "char",
+    "int8",
+    "uint8",
+    "int16",
+    "uint16",
+    "int32",
+    "uint32",
+    "int64",
+    "uint64",
+    "float32",
+    "float64",
+    "string",
+]
+INT_LENGTHS = [8, 16, 32, 64]
+FLOAT_LENGTHS = [32, 64]
 
-def _clean_name(name: str) -> str:
-    return re.sub(r"\W", "", name)
+UINT_REGEX = re.compile(r"^uint(\d+)")
+INT_REGEX = re.compile(r"^int(\d+)")
+FLOAT_REGEX = re.compile(r"^float(\d+)")
 
 
-def export(
-    current_pkg: information.DataPkg,
-    current_path: pathlib.Path,
-) -> None:
-    """Export a Capella data package to ROS messages."""
-    current_path.mkdir(parents=True, exist_ok=True)
-    for cls_obj in current_pkg.classes:
-        fields = []
-        for prop_obj in cls_obj.owned_properties:
-            try:
-                card = data_model.Range(
-                    prop_obj.min_card.value, prop_obj.max_card.value
-                )
-            except AttributeError:
-                card = data_model.Range("1", "1")
-            type_def = data_model.TypeDef(name=prop_obj.type.name, card=card)
-            prop_def = data_model.FieldDef(
-                type=type_def,
-                name=prop_obj.name,
-                description=prop_obj.description or "",
+class RosExportHelper:
+    @staticmethod
+    def int_bytes(length: int) -> int:
+        """Return ROS byte length for integers."""
+        for int_length in INT_LENGTHS:
+            if length <= int_length:
+                return int_length
+        raise ValueError(f"Invalid integer length {length}")
+
+    @staticmethod
+    def float_bytes(length: int) -> int:
+        """Return ROS byte length for floats."""
+        for float_length in FLOAT_LENGTHS:
+            if length <= float_length:
+                return float_length
+        raise ValueError(f"Invalid float length {length}")
+
+    @staticmethod
+    def make_doc_str(markup: str) -> list[str]:
+        _parser = LineSeparationHTMLParser()
+        _parser.feed(markup)
+
+        return _parser.text_list
+
+    @staticmethod
+    def make_snake_case(name: str) -> str:
+        """Convert all cases to snake_case."""
+        name = re.sub("([a-z0-9])([A-Z])", r"\1_\2", name)
+        name = re.sub("(.)([A-Z][a-z]+)", r"\1_\2", name)
+        name = name.lower()
+        # Replace invalid characters with underscores
+        name = re.sub("[^a-z0-9_]", "_", name)
+        name = re.sub("^[^a-z]+", "", name)
+        name = re.sub("_+", "_", name)
+        return re.sub("_$", "", name)
```

Comment:
This can lead to empty names:

```python
from capella_ros_tools.exporter import RosExportHelper as e
print(repr(e.make_snake_case("1+1=2"))) # ''
print(repr(e.make_snake_case("A 1+1=2"))) # 'a_1_1_2'
```